packages/jss/src/index.d.ts

```diff
@@ -7,35 +7,38 @@ import {Properties as CSSProperties} from 'csstype'
 // TODO: refactor to only include Observable types if plugin is installed.
 import {Observable} from 'indefinite-observable'
 
-// TODO: Type data better, currently typed as any for allowing to override it
-type Func<R> = ((data: any) => R)
+type Func<P, T, R> = T extends undefined ? ((data: P) => R) : ((data: P & {theme: T}) => R)
 
-type NormalCssProperties = CSSProperties<string | number>
+type NormalCssProperties = css.Properties<string | number>
 type NormalCssValues<K> = K extends keyof NormalCssProperties
-  ? NormalCssProperties[K]
+  ? NormalCssProperties[K] | JssValue
   : JssValue
 
-export type JssStyle =
+export type JssStyle<Props = any, Theme = undefined> =
   | {
       [K in keyof NormalCssProperties]:
         | NormalCssValues<K>
-        | JssStyle
-        | Func<NormalCssValues<K> | JssStyle | undefined>
+        | JssStyle<Props, Theme>
+        | Func<Props, Theme, NormalCssValues<K> | JssStyle<undefined, undefined> | undefined>
         | Observable<NormalCssValues<K> | JssStyle | undefined>
     }
   | {
       [K: string]:
         | JssValue
-        | JssStyle
-        | Func<JssValue | JssStyle | undefined>
+        | JssStyle<Props, Theme>
+        | Func<Props, Theme, JssValue | JssStyle<undefined, undefined> | undefined>
         | Observable<JssValue | JssStyle | undefined>
     }
 
-export type Styles<Name extends string | number | symbol = string> = Record<
+export type Styles<
+  Name extends string | number | symbol = string,
```

Comment:
Is Name necessary?

Reply:
I think this should not be generic, we could add it to Record directly.

Reply:
Hey @hosseinmd!

Yes, `Name` is necessary. Without it, users would be unable to guarantee intellisense on class names.

Additionally, `Styles` must remain a generic type in order for users to be able to specify `Props` and `Theme`.

Reply:
I think I misunderstood part of your question. I'll investigate when I get some time.

Reply:
I think you are right